src/ruby/cli.rs

```diff
@@ -0,0 +1,153 @@
+    // `programfile`. All trailing arguments are ARGV to the script.
+    //
+    // If there are `-e` arguments given, there is no programfile and all
+    // positional arguments are ARGV to the inline script.
+    //
+    // ```console
+    // $ ruby -e 'puts ARGV.inspect' a b c
+    // ["a", "b", "c"]
+    // $ cat foo.rb
+    // puts ARGV.inspect
+    // $ ruby foo.rb a b c
+    // ["a", "b", "c"]
+    // $ ruby bar.rb a b c
+    // ruby: No such file or directory -- bar.rb (LoadError)
+    // ```
+    if commands.is_empty() {
+        if let Some(programfile) = matches.get_one::<PathBuf>("programfile").cloned() {
+            args = args.with_programfile(Some(programfile));
+            if let Some(argv) = matches.get_many::<OsString>("arguments") {
+                let ruby_program_argv = argv.map(Clone::clone).collect::<Vec<_>>();
+                args = args.with_argv(ruby_program_argv);
+            }
+        }
+    } else {
+        args = args.with_commands(commands);
+        if let Some(first_arg) = matches.get_one::<PathBuf>("programfile").cloned() {
+            if let Some(argv) = matches.get_many::<OsString>("arguments") {
+                let ruby_program_argv = [OsString::from(first_arg)]
+                    .into_iter()
+                    .chain(argv.map(Clone::clone))
+                    .collect::<Vec<_>>();
+                args = args.with_argv(ruby_program_argv);
+            } else {
+                args = args.with_argv(vec![OsString::from(first_arg)]);
+            }
+        }
+    }
+
+    Ok(args)
+}
+
+// NOTE: This routine is plucked from `ripgrep` as of commit
+// `9f924ee187d4c62aa6ebe4903d0cfc6507a5adb5`.
+//
+// `ripgrep` is licensed with the MIT License Copyright (c) 2015 Andrew Gallant.
+//
+// https://github.com/BurntSushi/ripgrep/blob/9f924ee187d4c62aa6ebe4903d0cfc6507a5adb5/LICENSE-MIT
+//
+// See https://github.com/artichoke/artichoke/issues/1301.
+
+/// Returns a clap matches object if the given arguments parse successfully.
+///
+/// Otherwise, if an error occurred, then it is returned unless the error
+/// corresponds to a `--help` or `--version` request. In which case, the
+/// corresponding output is printed and the current process is exited
+/// successfully.
+fn clap_matches<I, T>(args: I) -> Result<ArgMatches, Box<dyn Error + Send + Sync + 'static>>
+where
+    I: IntoIterator<Item = T>,
+    T: Into<OsString> + Clone,
+{
+    let err = match cli().try_get_matches_from(args) {
+        Ok(matches) => return Ok(matches),
+        Err(err) => err,
+    };
+    if err.use_stderr() {
+        return Err(err.into());
+    }
+    // Explicitly ignore any error returned by write!. The most likely error
+    // at this point is a broken pipe error, in which case, we want to ignore
+    // it and exit quietly.
+    //
+    // (This is the point of this helper function. clap's functionality for
+    // doing this will panic on a broken pipe error.)
```

Comment:
I thought the broken pipe issue has been fixed for a while...

Reply:
I think it has been, but this code has worked the whole time and doesn't have that much maintenance cost for me to carry. I also prefer having visibility into all calls to `process::exit` instead of having them happen in a library if I can help it.

Reply:
But I can definitely clean up this comment since the clap side of things has been fixed.

Reply:
I just pushed a commit that removes this parenthetical comment. Thanks for catching that!

Reply:
You can also use `err.print()` so you can control the exiting but get clap to do the printing, including colored output.  If you want more control over that and are fine with ANSI escape codes, you can now also call `err.render().ansi()`.

Reply:
Oh that's awesome!

## Before

<img width="779" alt="Screen Shot 2022-09-28 at 8 23 53 PM" src="https://user-images.githubusercontent.com/860434/192931725-35772305-9ba1-4e00-a051-e74bf1e06d4f.png">

## After

<img width="1042" alt="Screen Shot 2022-09-28 at 8 24 06 PM" src="https://user-images.githubusercontent.com/860434/192931703-cab299da-5157-4226-90f2-69b66a4ff25a.png">

Reply:
Glad to be of help!